### PR TITLE
support ccache out of the box

### PR DESCRIPTION
**Pull request by rongou**
This adds support for `ccache` out of the box without asking the developer to do anything on their end, at the same time still allows the flexibility to configure their own setup (e.g. using both `ccache` and `distcc`).
```console
$ ccache -C  # clear the cache
$ time make -j12
...
real	48m36.971s
user	430m51.460s
sys	11m54.713s

$ make clean  # now everything is cached
$ time make -j12
...
real	0m12.049s
user	1m4.590s
sys	0m16.951s
```

### CODE REVIEW

File: cpp/cmake/Modules/ccache.cmake [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Use ccache if possible.
```
Comment (kkraus14) [inline cpp/cmake/Modules/ccache.cmake:1]: This file needs the NVIDIA copyright at the top of it

**Review by kkraus14 — CHANGES_REQUESTED** [review]
one copyright issue, otherwise it LGTM

**Comment by harrism** [comment]
I would like @trxcllnt to review in the context of rapids-compose, since many of us depend on it and I wouldn't want to disrupt a lot of developers.

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -22,6 +22,8 @@ if(NOT CMAKE_CUDA_COMPILER)
   message(SEND_ERROR "CMake cannot locate a CUDA compiler")
 endif(NOT CMAKE_CUDA_COMPILER)
 
+include(cmake/Modules/ccache.cmake)
```
Comment (trxcllnt) [inline cpp/CMakeLists.txt:25]: Can this be guarded by a CMake option? Changing the CMake compiler launcher will break anyone using ccache compiler rsymlinks.

**Review by trxcllnt — CHANGES_REQUESTED** [review]

File: cpp/cmake/Modules/ccache.cmake (hunk 1 above)
Comment (rongou) [reply cpp/cmake/Modules/ccache.cmake:17]: Done.

**Review by dillon-cullinan — APPROVED** [review]

**Review by mike-wendt — CHANGES_REQUESTED** [review]
In our images we use our own version of `ccache` for RAPIDS images. `gpuci-ccache` with this [feedstock](https://github.com/rapidsai/ccache-feedstock/blob/master/recipe/meta.yaml) and it is admittedly outdated. My concern with depending on the upstream version is we've had a lot of issues with breaking changes that impacts our builds. The goal is to support `ccache` in gpuCI builds during this release so we need to align on our use here as we can't be using incompatible versions.

The [scout](https://github.com/rapidsai/scout] project was started to centralize our `ccache` work. It has been dormant but with Project Flash rolling out we're starting work on this again. Before this or the counterpart to this rapidsai/rmm#657 is merged, can we please setup a call to discuss this more broadly and align over the next day or two before the holidays start?

File: cpp/CMakeLists.txt (hunk 2 above)
Comment (rongou) [reply cpp/CMakeLists.txt:25]: Done.

**Comment by rongou** [comment]
@mike-wendt @trxcllnt made the changes as discussed offline. PTAL

**Comment by rongou** [comment]
Hmm, I think I might have a fix for the nvcc symlink issue. Hold on.

**Comment by rongou** [comment]
Closing this for now. Will rely on https://gitlab.kitware.com/cmake/cmake/-/merge_requests/5620.

**Pull request closed without merging**